### PR TITLE
Verification options

### PR DESCRIPTION
**Pull request by loosebazooka**
This makes it easier to verify things like github idtokens that change the SAN on each release tag.

~~needs #730 and #731~~

Turns out there was OID processing that changed because old fulcio OIDs were raw strings and new ones are DER Encoded UTF8. see: https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#extension-values

### CODE REVIEW

File: sigstore-java/src/main/java/dev/sigstore/VerificationOptions.java [hunk 1]
```diff
@@ -27,17 +29,31 @@ public interface VerificationOptions {
...
 
-    Map<String, String> getOther();
+    Map<String, StringMatchable> getOther();
 
     static ImmutableCertificateIdentity.Builder builder() {
       return ImmutableCertificateIdentity.builder();
     }
   }
 
+  interface StringMatchable {
```
Comment (loosebazooka) [inline sigstore-java/src/main/java/dev/sigstore/VerificationOptions.java:43]: this is the interface that allows strings or regexs

**Comment by loosebazooka** [comment]
This is ready but I'm going to pull some parts out so it's easier to review.

File: sigstore-java/src/main/java/dev/sigstore/fulcio/client/FulcioCertificateMatcher.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  /** For OIDs with DER encoded UTF8 string entries */
+  public abstract Map<String, StringMatcher> getOidDerUtf8Strings();
+
+  /**
+   * For comparing raw bytes of the full ASN.1 object extension value as defined by EXTENSION in <a
+   * href="https://www.rfc-editor.org/rfc/rfc5280#section-4.1">rfc5280</a>
+   */
+  public abstract Map<String, byte[]> getOidBytes();
+
+  private static final Logger log = Logger.getLogger(FulcioCertificateMatcher.class.getName());
```
Comment (loosebazooka) [inline sigstore-java/src/main/java/dev/sigstore/fulcio/client/FulcioCertificateMatcher.java:56]: Most of what's below this line was pulled directly from the old FulcioCertificateVerifier class, now that we've kinda standardized on matcher

**Comment by loosebazooka** [comment]
kk, this should be ready to go @vlsi @patflynn

File: sigstore-java/src/main/java/dev/sigstore/KeylessVerifier.java [hunk 3]
```diff
@@ -208,4 +206,18 @@ public void verify(byte[] artifactDigest, Bundle bundle, VerificationOptions opt
...
     }
   }
+
+  @VisibleForTesting
+  void checkCertificateMatchers(X509Certificate cert, List<CertificateMatcher> matchers)
+      throws KeylessVerificationException {
+    try {
+      if (matchers.size() > 0 && matchers.stream().noneMatch(matcher -> matcher.test(cert))) {
+        throw new KeylessVerificationException(
+            "No provided certificate identities matched values in certificate");
```
Comment (patflynn) [inline sigstore-java/src/main/java/dev/sigstore/KeylessVerifier.java:216]: would it be helpful here to include matcher spec in the exception message?

File: sigstore-java/src/main/java/dev/sigstore/VerificationOptions.java [hunk 4]
```diff
@@ -15,26 +15,38 @@
...
-    String getIssuer();
-
-    String getSubjectAlternativeName();
-
-    Map<String, String> getOther();
+  List<CertificateMatcher> getCertificateMatchers();
+
+  /**
+   * An interface for allowing matching of certificates. Use {@link #fulcio()} to instantiate the
+   * default {@link FulcioCertificateMatcher} implementatio . Custom implementations may throw
```
Comment (patflynn) [inline sigstore-java/src/main/java/dev/sigstore/VerificationOptions.java:33]:
```suggestion
   * default {@link FulcioCertificateMatcher} implementation. Custom implementations may throw
```

File: sigstore-java/src/main/java/dev/sigstore/fulcio/client/FulcioCertificateMatcher.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+  public abstract Map<String, StringMatcher> getOidRawStrings();
+
+  /** For OIDs with DER encoded UTF8 string entries */
+  public abstract Map<String, StringMatcher> getOidDerUtf8Strings();
+
+  /**
+   * For comparing raw bytes of the full ASN.1 object extension value as defined by EXTENSION in <a
+   * href="https://www.rfc-editor.org/rfc/rfc5280#section-4.1">rfc5280</a>
+   */
+  public abstract Map<String, byte[]> getOidBytes();
```
Comment (patflynn) [inline sigstore-java/src/main/java/dev/sigstore/fulcio/client/FulcioCertificateMatcher.java:54]: maybe document why it's a map and what the keys are?

File: sigstore-java/src/main/java/dev/sigstore/KeylessVerifier.java (hunk 3 above)
Comment (loosebazooka) [reply sigstore-java/src/main/java/dev/sigstore/KeylessVerifier.java:216]: yeah probs

File: sigstore-java/src/main/java/dev/sigstore/fulcio/client/FulcioCertificateMatcher.java (hunk 5 above)
Comment (loosebazooka) [reply sigstore-java/src/main/java/dev/sigstore/fulcio/client/FulcioCertificateMatcher.java:60]: done

File: sigstore-java/src/main/java/dev/sigstore/KeylessVerifier.java (hunk 3 above)
Comment (loosebazooka) [reply sigstore-java/src/main/java/dev/sigstore/KeylessVerifier.java]: done

**Review by patflynn — APPROVED** [review]

**Pull request merged**